### ISSUE TITLE
 MXE + Qt + CMake = linking error (lcms2) 

### ISSUE THREAD

**Issue by starius** [mxe/mxe#852]
Qt [can be used](http://www.cmake.org/Wiki/CMake/Tutorials/Qt) with cmake as well as with qmake.

Using cmake with Qt, I get linking error because of missing `-llcms2`. I created minimal example reproducing the error as a patch for MXE. Apply  pull request #851 and [the patch](https://gist.github.com/starius/3dc64ab7c23c4457fd98) and type `make qt`. It builds Qt example with cmake.

I get the following errors:

```
Linking CXX executable qt-test-cmake.exe
/usr/bin/cmake -E cmake_link_script CMakeFiles/qt-test-cmake.dir/link.txt --verbose=1
/usr/bin/cmake -E remove -f CMakeFiles/qt-test-cmake.dir/objects.a
/home/mxe/mxe-conf-cmake/usr/bin/x86_64-w64-mingw32.static-ar cr CMakeFiles/qt-test-cmake.dir/objects.a @CMakeFiles/qt-test-cmake.dir/objects1.rsp
/home/mxe/mxe-conf-cmake/usr/bin/x86_64-w64-mingw32.static-g++  -O3 -DNDEBUG    -Wl,--whole-archive CMakeFiles/qt-test-cmake.dir/objects.a -Wl,--no-whole-archive  -o qt-t
est-cmake.exe -Wl,--out-implib,libqt-test-cmake.dll.a -Wl,--major-image-version,0,--minor-image-version,0  /home/mxe/mxe-conf-cmake/usr/x86_64-w64-mingw32.static/qt/lib/l
ibQtGui.a -Wl,-Bstatic -lpng -ljpeg -Wl,-Bdynamic -ltiff -llzma -ljpeg -lz -Wl,-Bstatic -lmng -Wl,-Bdynamic -limm32 -lwinmm /home/mxe/mxe-conf-cmake/usr/x86_64-w64-mingw3
2.static/qt/lib/libQtCore.a -Wl,-Bstatic -lz -Wl,-Bdynamic -lws2_32 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 
/home/mxe/mxe-conf-cmake/usr/lib/gcc/x86_64-w64-mingw32.static/5.2.0/../../../../x86_64-w64-mingw32.static/lib/../lib/libmng.a(libmng_cms.o): In function `mng_correct_ful
l_cms':
/home/mxe/mxe-conf-cmake/tmp-libmng-x86_64-w64-mingw32.static/libmng-2.0.3/libmng_cms.c:448: undefined reference to `cmsDoTransform'
/home/mxe/mxe-conf-cmake/usr/lib/gcc/x86_64-w64-mingw32.static/5.2.0/../../../../x86_64-w64-mingw32.static/lib/../lib/libmng.a(libmng_cms.o): In function `mnglcms_creates
rgbprofile':
/home/mxe/mxe-conf-cmake/tmp-libmng-x86_64-w64-mingw32.static/libmng-2.0.3/libmng_cms.c:148: undefined reference to `cmsWhitePointFromTemp'
/home/mxe/mxe-conf-cmake/tmp-libmng-x86_64-w64-mingw32.static/libmng-2.0.3/libmng_cms.c:154: undefined reference to `cmsBuildGamma'
...
```

After uncommenting this line:

``` cmake
target_link_libraries(qt-test-cmake lcms2)
```

the error was fixed.

Probably cmake or pkg-config doesn't know about libmng's dependency on lcms2.

_Another issue_. If I add `cmake_policy(SET CMP0017 NEW)` to fix cmake warnings (see #800), I get linking errors because of missing `-llzma`:

```
/home/mxe/usr/lib/gcc/i686-w64-mingw32.static/5.2.0/../../../../i686-w64-mingw32.static/lib/../lib/libtiff.a(tif_lzma.o): In function `LZMASetupEncode':
/home/mxe/tmp-tiff-i686-w64-mingw32.static/tiff-4.0.5/libtiff/tif_lzma.c:236: undefined reference to `lzma_end'
```

**Comment by tonytheodore** [comment]
> If I add cmake_policy(SET CMP0017 NEW)

I'd recommend setting that, it seems to do the right thing.

> Probably cmake or pkg-config doesn't know about libmng's dependency on lcms2.

Try this:
```
find_package(PkgConfig REQUIRED)
pkg_check_modules(LIBMNG libmng)
...
target_link_libraries(qt-test-cmake ${QT_LIBRARIES} ${LIBMNG_LIBRARIES})
```

**Comment by starius** [comment]
>find_package(PkgConfig REQUIRED)
>pkg_check_modules(LIBMNG libmng)
>...
>target_link_libraries(qt-test-cmake ${QT_LIBRARIES} ${LIBMNG_LIBRARIES})

It does not fix linking errors. Link command is as follows:
```
/home/mxe/mxe-conf-cmake/usr/bin/x86_64-w64-mingw32.static-g++  -O3 -DNDEBUG    -Wl,--whole-archive CMakeFiles/qt-test-cmake.dir/objects.a -Wl,--no-whole-archive  -o qt-test-cmake.exe -Wl,--out-implib,libqt-test-cmake.dll.a -Wl,--major-image-version,0,--minor-image-version,0  /home/mxe/mxe-conf-cmake/usr/x86_64-w64-mingw32.static/qt/lib/libQtGui.a -Wl,-Bstatic -lpng -ljpeg -Wl,-Bdynamic -ltiff -llzma -ljpeg -lz -Wl,-Bstatic -lmng -Wl,-Bdynamic -limm32 -lwinmm /home/mxe/mxe-conf-cmake/usr/x86_64-w64-mingw32.static/qt/lib/libQtCore.a -Wl,-Bstatic -lz -Wl,-Bdynamic -lws2_32 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 
```

LIBMNG is a Qt's dependency, that is why I want cmake to do all this stuff automatically. If I included this to CMakeLists.txt of a real project, I would have to add `IF (MINGW)` not to break build on other platforms...

**Comment by tonytheodore** [comment]
As some stage in the past, this [CMakeLists.txt](https://gist.github.com/tonytheodore/c0a5e496cf79ffac2718#file-qt-test-cmakelists-txt) worked with [this invocation](https://gist.github.com/tonytheodore/c0a5e496cf79ffac2718#file-qt-mk). Using `Qt4::QtGui` (or `Qt5::Widgets`) instead of `${QT_LIBRARIES}` seemed to find the all the right libs. Unfortunately, it doesn't work any longer.

**Comment by tonytheodore** [comment]
Hmm... it's likely that only worked with shared builds - I might have committed it otherwise. Searching around for cmake and qt static builds seems to lead towards parsing the *.prl files that qmake uses. 

There's a [wrapper script for qt4](http://lists.gnu.org/archive/html/mingw-cross-env-list/2010-11/msg00069.html) and it looks like functionality was [added to qtbase](http://code.qt.io/cgit/qt/qtbase.git/commit/mkspecs/cmake/Qt5BasicConfig.cmake.in?id=d4835da66f7970877c6a677845b41ad1c1045077), but [subsequently removed](http://code.qt.io/cgit/qt/qtbase.git/commit/mkspecs/cmake/Qt5BasicConfig.cmake.in?id=102e1822ffcdc9954d3c698f863734a8083e349c). 

At this stage, I'd say it's unlikely that cmake can do this automatically.

**Comment by tonytheodore** [comment]
Apropos, a workaround for vtk6 in #857. Is the real issue something to do with our libmng?

**Comment by muellni** [comment]
> Is the real issue something to do with our libmng?

No, the problem is in the CMake host installation, namely the file Qt4ConfigDependentSettings.cmake (/usr/share/cmake-3.3/Modules/Qt4ConfigDependentSettings.cmake on my machine). 
It contains
```
if(QT_QCONFIG MATCHES "system-mng")
  find_library(MNG_LIBRARY NAMES mng)
  set(QT_QTGUI_LIB_DEPENDENCIES ${QT_QTGUI_LIB_DEPENDENCIES} ${MNG_LIBRARY})
endif()
```
which completely misses libmngs dependencies.
I hacked this by setting ```MNG_LIBRARY``` to ```mng lcms2```.

I still think we should provide a MXE version of CMake [as proposed some time] (https://lists.gnu.org/archive/html/mingw-cross-env-list/2011-12/msg00012.html) ago. This would give us a place to collect static linking patches. 
MXE already has a CMake package, so maybe we should use it and provide a way to use it for MXE users. But the current ```src/cmake``` directory is a good start.

**Comment by tonytheodore** [comment]
> I still think we should provide a MXE version of CMake as proposed some time ago

Agreed. The bitbucket links are broken, but the intent is exactly what we're trying to achieve. We've always built `qmake`, have been supplying a custom `pkgconf` to replace `pkg-config` for some time, and `cmake` builds easily on all platforms I've tested. 

I'm not sure if your proposal actually requires 3.3, but feel free to use any new features that make it simpler. If Kitware projects still need patching...

**Comment by tonytheodore** [comment]
I've [added a simple cmake test to qt](https://github.com/mxe/mxe/commit/dd34bab1b47e929b7becb5f1fd9cee9696cef44b), but the same doesn't work with Qt5.

**Comment by tonytheodore** [comment]
> I'm not sure if your proposal actually requires 3.3

`cgal` and `vmime` break with v3.3.2 but I'm inclined to use the latest version anyway; both for proposing patches upstream and using features like the `AUTOMOC` etc. It may not be possible to support a range that spans 2.8.9 on Wheezy through 3.3.

**Comment by tonytheodore** [comment]
This specific issue was fixed in #857 and #1645 has a general solution.
